### PR TITLE
Added forage to remedies for workorders for recipes without herbs in Society.

### PR DESCRIPTION
**Pull request by Dartellum**
This update will try to collect the herb for a recipe that is deemed false for workorders. Workorders are set to false if the herb is not purchasable. This has caused many to stop progressing to high skill as the higher Tiers require herbs one cannot purchase from the Societies. 

~~*************** This change is not ready for pushing! *************************~~

Putting here in case something happens outside my control. ~~I need to check the tool needed for sap and moss. Also want to change the walk_to to the stored room in base-recipes but need to confirm the press and grinder are in the same variable.~~ Need to add
```
     press-grinder-rooms:
    - xxxx
    - xxxx
    idle-room: xxxx
```
for Muspar'i, Ratha, and Fang Cove. A shame `lrf  12372` or any room already defined does not give press and grinder information. I can get Muspar'i and Ratha unless someone beats me to it. I do not believe I can get to Fang Cove. The idle room is just a room with no press and grinder. 

Adds the following variables to `base`:

```
# container used to hold herbs for remedy.lic (optional) Not optional if overriding
herb_container:
# Following settings, including herb_container:, are required to override workorders for remedies
# container used for raw herbs foraged - used by alchemy.lic script
alchemy_herb_storage:
# quantity of gathered herbs, recall a workorder requires 25 herbs per finished product
# Set to your workorder_max_items: setting times 4. Example, if workorder_max_items: 4,
# set this setting to 100.
alchemy_herb_quantity:
# How many prepared herbs you want:
alchemy_prep_quantity:
# if you want to forage careful
alchemy_forage_type:
# Set an override for foraging if Outdoorsmanship is high enough to find the herb anywhere
forage_override_room:
# Set an override for town
forage_override_town:
# Set this to true if you want to always forage for your herbs for workorders
workorders_override_store: false
```

;alchemy script is usable standalone and with craft. It has a craft function as well. Can work on the craft script passing in the recipe based on ranks after obtaining and processing the herb(s). See examples below.

;alchemy usage: 
;alchemy aloe forage - forages for aloe leaves based on settings in base-ingredients or overridden by settings in yaml. 
;alchemy aloe prepare - will walk to location based on hometown or override to process herbs in Society
;alchemy "skin ungent" craft - will craft skin ungent - calls remedy with all the parameters
;alchemy aloe forage 50 - will check and then forage until you have 50 items of the herb
Add debug to the end of it for some diagnostic output.

;craft scipt can determine ranks and recipe and pass those parameters into alchemy to forage, process, and craft. Making one you need 25 herbs.

if Alchemy skill <= <some number>
wait_for_script_to_complete('alchemy',[herb, 'forage', 25])
wait_for_script_to_complete('alchemy',[herb, 'prepare'])
~~wait_for_script_to_complete('alchemy',["skin ungent", 'craft'])~~

Finally, ;workorders remedies will figure everything out. Two options, one it will only override the recipe if you have it explicitly in your yaml under **workorder_recipes:** like this:
```
workorder_recipes:
  remedies:
  - some skin ungent
  #- some skin tonic

```
You will only look for those recipes. A nice bonus is it will forage and process for recipes with an empty herb1_stock or herb2_stock. The second option is to override always to gather your own herbs with **workorders_override_store: false**. This option does not need the specific remedies listed under **workorder_recipes:**.

### CODE REVIEW

**Comment by jayRyan24** [comment]
> 
> *************** This change is not ready for pushing! *************************
> 

You can remove the above, and just convert this to a draft. On the top right of the PR, you'll see:

image

There ya go!

File: workorders.lic [hunk 1]
```diff
@@ -489,6 +489,55 @@ class WorkOrders
     end
   end
 
+  def gather_process_herb(herb, herb_to_purchase)
```
Comment (jayRyan24) [inline workorders.lic:492]: What do you think about breaking this into a forage.lic script? I could find that useful outside of the workorders as well.

`;forage <quantity> <herb name> <rnum>`

**Review by jayRyan24 — CHANGES_REQUESTED** [review]
Overall I think this is a great start, but shifting this forage script to a standalone PR, and then utilizing it within workorders seems like a better course.

**Comment by jayRyan24** [comment]
It would actually be really cool if this script took two sets of parameters. 

**#1**
`;forage <quantity> <herb name> <rnum>
- This checks the current quantity against the quantity submitted, and then goes to the room number and farms.

**#2**
`;forage stock`
- This checks against a set of YAML settings and does the same as above, but goes through them and farms what it needs. 

```
forage_stock:
  herb1:
    quantity: 50
    room: 12345
  herb2:
    quantity: 25
    room: 1345
  herb3:
    quantity: 10
    room: 5683
```

**Comment by Dartellum** [comment]
Powers, I think, put together a script that will do what you ask. I have the two parts two it, alchemy and base-ingredients.yaml. It does exactly what you are asking. I know of a few others that may have it but it is not widely used. I can add those pieces to this and incorporate it into workorders. I will add tomorrow and start the testing.

**Comment by jayRyan24** [comment]
I am not sure it should be incorporated into workorders as much as it needs to be a separate script which does a task, and workorders can utilize it.

**Comment by jayRyan24** [comment]
> Powers, I think, put together a script that will do what you ask. I have the two parts two it, alchemy and base-ingredients.yaml. It does exactly what you are asking. I know of a few others that may have it but it is not widely used. I can add those pieces to this and incorporate it into workorders. I will add tomorrow and start the testing.

Sounds good, but I am not sure it should be incorporated into workorders as much as it needs to be a separate script which does a task, and workorders can utilize it.

**Comment by Dartellum** [comment]
That is what I meant, incorporate workorders calling it. I need to make sure the script works and it needs some more data points in the base-ingredients.yaml file. But, once done, workorders would pass the herb required to it, if it is not on hand already, and it will do the gathering and processing instead of inside workorders and then resume workorders.

File: profiles/base.yaml [hunk 2]
```diff
@@ -114,6 +114,14 @@ fatigue_regen_threshold: 90
...
 # Minimum balance to perform certain actions (currently only whirlwind)
 balance_regen_threshold: 8 #solidly balanced
+# Pattern Hues settings
+# gesture is either random, custom, or a specific setting
+pattern_hues:
+  duration: 1500
+  gesture: random
+  pattern_hues_styles:
+    pattern:
+    hue:
```
Comment (Dartellum) [inline profiles/base.yaml:124]: Started this PR before this was merged. I will fix it once ready to go.

File: profiles/base.yaml [hunk 3]
```diff
@@ -1511,3 +1536,11 @@ faux_atmo_no_use_rooms:
...
 # Will attempt to invoke a tessera during crossing-training between activities, 10 minute cooldown.
 use_tessera_during_crossing_training: false
+
+# Feed-cloak settings
+# If your cloak can't feed in your current room
+# then feed-cloak will move to this specified room id.
+# If not specified, or if the cloak can't feed here either,
+# then the script will move to your `safe_room`.
+# Example: 992 # Crossing, [Northeast Wilds, Outside Northeast Gate]
+feed_cloak_room:
```
Comment (Dartellum) [inline profiles/base.yaml:1546]: Started this PR before this was merged. I will fix it once ready to go.

File: remedy.lic [hunk 4]
```diff
@@ -177,7 +177,8 @@ class Remedy
       @herb1 = 'dried flowers' unless exists?('red flowers') || exists?('blue flowers')
     end
     get_item(@container)
-    bput("lower #{@container} ground", 'You lower') # lower to count and combine herbs
+
+    bput("lower my #{@container} ground", 'You lower') # lower to count and combine herbs
```
Comment (Dartellum) [inline remedy.lic:181]: Started this PR before this was merged. I will fix it once ready to go.

File: workorders.lic (hunk 1 above)
Comment (Dartellum) [reply workorders.lic]: See updated comment.

**Comment by Dartellum** [comment]
I will resolve the conflict once ready to go. Weird it is the same setting in the same place but it sees it as a conflict.

File: alchemy.lic [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        { name: 'prepare', regex: /prepare/i, optional: true },
+        { name: 'craft', regex: /craft/i, optional: true },
+        { name: 'herb_quantity_needed', regex: /\d+/, optional: true, description: 'Used to calculate required herbs to forage and process.' },
+        { name: 'debug', regex: /debug/i, optional: true }
+      ]
+    ]
+
+    args = parse_args(arg_definitions)
+
+    echo "*** incorrect arguments ***" if args.craft and (args.forage or args.prepare)
```
Comment (rpherbig) [inline alchemy.lic:28]: 💡 Prefer `&&` and `||` to `and` and `or`

**Comment by rpherbig** [comment]
This is still marked as a draft

**Comment by rpherbig** [comment]
Please resolve the merge conflicts, this is a bit hard to review

File: alchemy.lic [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+      fput("put my #{ingredient['name']} in my #{@alchemy_herb_storage}")
+      echo quantity_needed if @debug
+    end
+  end
+
+  def prepare_ingredient(ingredient)
+    #dry? crush?
+    stow_hands
+    exit if @alchemy_prep_quantity <= count(ingredient['output'],@herb_container)
+    find_press_grinder_room(@press_location)
```
Comment (rpherbig) [inline alchemy.lic:113]: 💡 `DRCC.` please

File: alchemy.lic [hunk 7]
```diff
@@ -0,0 +1,164 @@
+=begin
+  Documentation: https://elanthipedia.play.net/Lich_script_repository#-------------
```
Comment (rpherbig) [inline alchemy.lic:2]: 💡 I assume `-------------` should be `alchemy`?

File: alchemy.lic [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    @herb_quantity_needed = args.herb_quantity_needed.to_i
+
+    ingredient = get_data('ingredients').ingredients.find{ |ingredient| ingredient['name'] =~ /#{args.target}/i } unless args.craft
+    # Set town for foraging
+    @forage_location = @forage_override_town || @forage_override_room || ingredient[@hometown]
+    echo ("@forage_location = #{@forage_location}") if @debug
+    @press_location = @forage_override_town || ingredient[@hometown]
+    echo ("@press_location = #{@press_location}") if @debug
+    recipe = get_data('recipes').crafting_recipes.find { |recipe| recipe['name'] =~ /#{args.target}/i } if args.craft
+    forage_ingredient(ingredient) if args.forage
```
Comment (rpherbig) [inline alchemy.lic:50]: ❓ Something seems off here. This line can run if `args.craft` is present, yet `ingredient` will be `nil`. Is that intended?

File: alchemy.lic [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    @alchemy_prep_quantity = settings.alchemy_prep_quantity
+    @alchemy_forage_type = settings.alchemy_forage_type
+    @forage_override_room = settings.forage_override_room
+    @forage_override_town = settings.forage_override_town
+    @debug = args.debug
+    @herb_quantity_needed = args.herb_quantity_needed.to_i
+
+    ingredient = get_data('ingredients').ingredients.find{ |ingredient| ingredient['name'] =~ /#{args.target}/i } unless args.craft
+    # Set town for foraging
+    @forage_location = @forage_override_town || @forage_override_room || ingredient[@hometown]
```
Comment (rpherbig) [inline alchemy.lic:45]: ❓ Where is `@forage_location` used?

File: alchemy.lic (hunk 5 above)
Comment (Dartellum) [reply alchemy.lic:28]: Sorry, was Powers logic and was working so I just skipped over. Will change.

File: alchemy.lic (hunk 6 above)
Comment (Dartellum) [reply alchemy.lic:113]: I have common-crafting and include DRCC. Do I still need to prefix it with `DRCC.`?

File: alchemy.lic (hunk 8 above)
Comment (Dartellum) [reply alchemy.lic:50]: Yes, intended. You call the script twice; once to gather and process and once to craft, if using this script to craft.

**Comment by Dartellum** [comment]
> Please resolve the merge conflicts, this is a bit hard to review

The conflict is because I started before those other changes were made. I forgot how to pull from upstream and merge with my branch that contains the changes.

File: alchemy.lic (hunk 6 above)
Comment (rpherbig) [reply alchemy.lic:116]: Need to, no. But it makes it clear where functions are coming from, i.e. easier for new people to get the hang of what's going on.

File: alchemy.lic (hunk 8 above)
Comment (rpherbig) [reply alchemy.lic:53]: That seems to me like this should be two separate scripts then: one to gather & process and one to craft. What do you think?

Comment (Dartellum) [reply alchemy.lic:53]: Could just make it part of `;craft` I suppose. But then one could not run it standalone and select whatever recipe they want to make outside of training.

Comment (Dartellum) [reply alchemy.lic:53]: I like the all-in-one approach.

Comment (Dartellum) [reply alchemy.lic:53]: Removed craft and day/night check (not used presently).

**Comment by rpherbig** [comment]
@Dartellum is this ready to be merged?

**Comment by Dartellum** [comment]
> @Dartellum is this ready to be merged?

I need to grab two more press and grinder locations. I need someone with Fang Cove access to give the room number for it there and a room to idle in.

**Comment by Dartellum** [comment]
Can add additional information in another PR, such as Fang Cove's press and grinder rooms.

**Comment by Dartellum** [comment]
@rpherbig Ready to merge.

**Pull request merged**